diff --git a/hw/mcu/nordic/nrf52xxx/src/hal_i2c.c b/hw/mcu/nordic/nrf52xxx/src/hal_i2c.c
index e71c1d2..47acb84 100644
--- a/hw/mcu/nordic/nrf52xxx/src/hal_i2c.c
+++ b/hw/mcu/nordic/nrf52xxx/src/hal_i2c.c
@@ -510,6 +510,7 @@ hal_i2c_master_write(uint8_t i2c_num, struct hal_i2c_master_data *pdata,
     regs->SHORTS = 0;
 
     hal_i2c_trigger_start(regs, &regs->TASKS_STARTTX);
+    regs->TASKS_RESUME = 1;
 
     start = os_time_get();
     for (i = 0; i < pdata->len; i++) {
